### PR TITLE
Add L2SqrtExpanded support to ivf_flat ANN indices

### PR DESCRIPTION
**Pull request by benfred**

### CODE REVIEW

**Review by achirkin — CHANGES_REQUESTED** [review]
Thanks, couple of minor suggestions here

File: cpp/include/raft/spatial/knn/detail/ann_quantized.cuh [hunk 1]
```diff
@@ -122,6 +122,8 @@ void approx_knn_build_index(const handle_t& handle,
 
   if (ivf_ft_pams && (metric == raft::distance::DistanceType::L2Unexpanded ||
                       metric == raft::distance::DistanceType::L2Expanded ||
+                      metric == raft::distance::DistanceType::L2SqrtExpanded ||
+                      metric == raft::distance::DistanceType::L2SqrtUnexpanded ||
```
[the remark this answers is not in the document]
Comment (benfred) [reply cpp/include/raft/spatial/knn/detail/ann_quantized.cuh]: I've changed to use a switch statement like you suggested - .

Note that this check in ann_quantized is going to be deleted in the next week or so - I'm working on removing the faiss dependency from RAFT, and once https://github.com/rapidsai/raft/pull/1121 is merged we won't need this check anymore.

File: cpp/include/raft/spatial/knn/detail/ivf_flat_search.cuh [hunk 2]
```diff
@@ -1215,6 +1218,13 @@ void search_impl(const handle_t& handle,
                             stream,
                             search_mr);
   }
+
+  // post-process
+  if (index.metric() == raft::distance::DistanceType::L2SqrtExpanded ||
+      index.metric() == raft::distance::DistanceType::L2SqrtUnexpanded) {
+    raft::linalg::unaryOp<float>(
+      distances, distances, n_queries * k, raft::sqrt_op(), handle.get_stream());
+  }
```
[the remark this answers is not in the document]
Comment (benfred) [reply cpp/include/raft/spatial/knn/detail/ivf_flat_search.cuh]: Good idea! 

I took a pass at adding a lambda argument to do post processing to queue.store in the last commit - can you take a look?

File: cpp/include/raft/spatial/knn/detail/topk/warpsort_topk.cuh [hunk 3]
```diff
@@ -591,9 +592,10 @@ class block_sort {
   }
 
   /** Save the content by the pointer location. */
-  __device__ void store(T* out, IdxT* out_idx) const
+  template <typename Lambda = raft::identity_op>
+  __device__ void store(T* out, IdxT* out_idx, Lambda post_process = raft::identity_op()) const
```
Comment (achirkin) [inline cpp/include/raft/spatial/knn/detail/topk/warpsort_topk.cuh:596]: Just for the record: I'm currently refactoring select_k in https://github.com/rapidsai/raft/pull/1085 , so I'd need to catch up with these changes there. I'm also planning to change `T` and `IdxT` into `OutT` and `OutIdxT` to allow saving the results in a type different from the working types `T` and `IdxT`.

**Review by achirkin — APPROVED** [review]
Thanks, LGTM!

File: cpp/include/raft/spatial/knn/detail/ann_quantized.cuh (hunk 1 above)
[the remark this answers is not in the document]
Comment (achirkin) [reply cpp/include/raft/spatial/knn/detail/ann_quantized.cuh]: Thanks, looking forward to #1121 !

**Review by cjnolet — APPROVED** [review]

**Pull request merged**